rust/vapid/src/lib.rs

```diff
@@ -0,0 +1,482 @@
+//! VAPID auth support
+//!
+//! This library only supports the latest VAPID-draft-02+ specification.
+//!
+//! Example Use:
+//! ```
+//!
+//! // Create a key from an existing EC Private Key PEM file.
+//! // You can generate this with
+//! // Key::generate().to_pem("pem/file/path.pem")?;
+//! let my_key = Key::from_pem("pem/file/path.pem")?;
+//!
+//! // Construct the Claims hashmap
+//! let mut claims:HashMap<String, serde_json::Value> = HashMap::new();
+//! claims.insert(
+//!     String::from("sub"), serde_json::Value::from("mailto:bob@example.com")
+//! );
+//! // while `exp` can be filled in for you, `aud` should point to the net location of the
+//! // Push server you wish to talk to. (e.g. `https://push.services.mozilla.org`)
+//! // `aud` is optional for Mozilla, but may be required for GCM/FCM or other systems.
+//! claims.insert(
+//!     String::from("aud"), serde_json::Value::from("https://host.ext")
+//! );
+//!
+//! // The result will contain the `Authorization:` header. How you inject this into your
+//! // request is left as an exercise.
+//! let authorization_header = sign(key, &mut claims)?;
+//!
+//! ```
+
+extern crate asn1;
+extern crate base64;
+extern crate openssl;
+extern crate serde_json;
+extern crate time;
+
+use std::collections::HashMap;
+use std::fs::File;
+use std::io::Read;
+use std::io::Write;
+use std::path::Path;
+
+use openssl::bn::BigNum;
+use openssl::bn::BigNumContext;
+use openssl::hash::MessageDigest;
+use openssl::ec;
+use openssl::nid;
+use openssl::sign::{Signer, Verifier};
+use openssl::pkey;
+
+pub struct Key(ec::EcKey);
+
+/// a Key is a helper for creating or using a VAPID EC key.
+///
+/// Vapid Keys are always Prime256v1 EC keys.
+///
+#[allow(dead_code)]
+impl Key {
+    fn new() -> Key {
+        Key(ec::EcKey::from_curve_name(Key::name()).expect(
+            "EC Prime256v1 curve not supported",
+        ))
+    }
+
+    fn name() -> nid::Nid {
+        nid::X9_62_PRIME256V1
+    }
+
+    fn group() -> ec::EcGroup {
+        ec::EcGroup::from_curve_name(Key::name()).expect("EC Prime256v1 curve not supported")
+    }
+
+    /// Read a VAPID private key stored in `path`
+    pub fn from_pem(path: String) -> Result<Key, String> {
```

Comment:
A few comments about this:

* This function signature indicates that it's taking the `path` argument as an owned value (a `String`) vs a non-owned value (a `&str`). The implementation, however, doesn't actually need the owned aspect of the string here, it just needs `&str` in theory. In general you'd take borrowed values in Rust when you don't need ownership (things with a `&`) and owned values only when you need to.
* For representing paths we've got the `Path` type in the standard library (borrowed version) and `PathBuf` (owned version). You can think of `&Path` like `&str` and `PathBuf` like `String`. Unfortunately though we've only got string literal syntax, not path literal syntax :(.
* If you want to get *super* fancy, the idiomatic way to take a "path as an argument" is to do something like:

```rust
pub fn from_pem<P>(path: P) -> Result<...> 
    where P: AsRef<Path>,
{
    // ...
}
```

but again, that's (in my opinion) overkill for an example like this. The generic argument makes it maximally ergonomic to call where you can do `Key::from_pem("foo")` rather than `Key::from_pem("foo".to_string())` or `Key::from_pem(Path::new("foo"))`. Just figured I'd mention it!

Reply:
Awesome! I'm using Path in other places, so I'll make things consistent.